webhooks: Updated docs for some complex integrations.

eeshangarg opened the pull request:
@rishig, @timabbott: FYI :)

timabbott commented on zerver/webhooks/beanstalk/doc.md:
```diff
@@ -1,19 +1,31 @@
+1. Click on the **Add a webhook** button.
 
-5. {!git-append-branches.md!}
+1. Set **Name** to a name of your choice, such as `Zulip Webhook`.
```
Probably "Zulip" is clear enough for this sort of context.

timabbott commented on zerver/webhooks/codeship/doc.md:
```diff
@@ -3,19 +3,16 @@ your build statuses.
+1. Click on the **+ New Notification** button.
 
-    where `PROJECT_ID` is the ID of your project in Codeship.
+1. Click on **Webhook**.
```
Should this be just part of the previous step, maybe spelled as "Select Webhook"?

eeshangarg commented:
@timabbott, @rishig: Updated with more! Thanks! :)

eeshangarg commented:
@timabbott: Also, I implemented your feedback :)

rishig commented:
Thanks @eeshangarg!
* Bitbucket: Probably "Zulip" is a good title, rather than "Zulip webhook". I would also say "Choose Webhooks" rather than "choose webhooks on the left hand side", unless it is a confusing enough UI that that would be helpful. Also let's go with "Check the Active checkbox" instead of the current "Remember to check the" 
* Beanstalk doesn't look done; there should be only ~5 instructions? Also the URL looks different, is that still correct? 
We should also put this into using the macros of the others, if it is easy (i.e. first step is "create the stream you would like to use for XXX notifications ..."). If reusing the macro is hard, even just making it be the identical text will make it easier to maintain it later. 
* hm, similarly for beeminder; steps 4, 5, and the first half of 6 should be one step. The second half of 6 should be it's own step. 

Maybe let's write the style guide first, so we can make sure we're in sync with all the pieces of the style?

After seeing these, I'm also realizing that the filled in screenshot is better than an empty one. (Or at least, that an empty one looks sort of strange.) No need to go back and change the ones you've already done, but maybe let's rethink what we want to do here. 

One possibility is just to remove all the screenshots; that has the benefit of not needing to standardize them, and the screenshots never going out of date. Also possibly we can then instead highlight the "Congratulations" screenshot, which is an advertisement for the webhook, rather than a random screenshot of the service.

Another possibility is to add a red box or something around each of the boxes that need to be filled.

I'm leaning slightly towards removing them, but interested in yours and Tim's thoughts. 
My thought is that the main goal of this text is to encourage the user to get started with the process, and then provide enough assistance that they never get scared or stuck. And for that I'm not sure how much the screenshot helps. Tim: three to look at are Airbrake (no screenshot), Appfollow (screenshot works well), and Beanstalk (screenshot doesn't work well, imo).

Also lastly just want to check that you are using the zulip integrations email and password for each of these sites? That will make it easy to go through and update when we do a round of updating in the future.

timabbott commented:
I think removing them is fine unless there's a specific reason they're required (e.g. something is very hard to find).  In that case, we probably want a smaller screenshot, with perhaps some drawing on the image done to mark the thing that's hard to find.

rishig commented:
Great, sounds good to me. In that case we can also change our numbering a bit, so that "click Save" is in the same instruction as "fill out the stuff". 

So a typical doc, where nothing is complicated or needs an extra picture, might look like

1. from macro
2. from macro
3. Go to blah. Click blah, and then click blah. [i.e. navigational instructions]
4. Enter in the URL you .... Check the Active button ... Click Save. [i.e. fill out the form and click save instructions].

eeshangarg commented:
@timabbott, @rishig: Updated with all the requested changes. I wouldn't be surprised if this might need another couple of rounds of feedback, especially the Facebook doc is tricky! Looking forward to your feedback! Also, I think I'll reiterate on this set of commits before doing any more since this might still need changes! Thanks! :)

rishig commented:
great, thanks! A few thoughts:

*Beeminder*
* So far, we've put "Create a URL" and "Create the bot" in the same step. For beeminder, they are different steps. Possibly the beeminder route is the correct one (make them separate steps), but if we go that route we should update our style guide notes to say that. (Or is the beeminder one separate because the URL instructions are longer?)
* Step 4 starts with "Next, ..". We shouldn't start with transition words, since it makes it more likely to introduce bugs in macro refactoring. 
* Now that we don't have the image, I think we should use casing like "Deadline & Reminders" or "Save changes" in the situation were it is in all caps in the beeminder UI. The bold, all caps is too scary.
* This webhook seems to operate in a strange way on the zulip end. What happens if you specify a stream and also click "Make this goal private"? Probably the private thing should be explained in step 3. 

Other notes:
* The CircleCI instructions seem unfinished, but we can deal with that later
* If there is ending punctuation in the setting, like "Enable this checkbox?", we should leave the punctuation out of our instructions. Starting punctuation, like "+ New Notification" should stay. 
* If there are 4 sentences, we should generally join the middle ones with an "and". E.g. instead of "Click W. Click X. Click Y. Click Z.", "Click W. Click X, and click Y. Click Z". 
* We should generally not include "log on to your ..". 

This is getting to the point of being pretty nit-picky, though I feel like we're making good progress towards a template that we can just apply to most new integrations without thinking.

I think next steps are probably to:
* Start writing things down into a style guide. A paper doc seems like a good place to start; I can create one if you'd like.
* Get 5 of the easiest ones merged from this list, so we reduce the number of integrations we're looking at.

The facebook one is a bit complicated, so we can do that after we nail the simple ones. But at a first pass:
* We don't need the notice at the bottom ("This integration is not created by ...")
* Do we need a token? If not, we should just set it to something reasonable, like `token=unused`.
* "Select the User category .." should be the start of a new step. Also, we can remove the "therefore".
* Replace "Zulip server" with "Zulip organization" in the last row. 

Thanks!

eeshangarg commented:
@rishig: I have a few thoughts:

#### About Facebook:

Implemented everything except I don't agree that we should just set `token` to a value such as `unused`. The `token` argument is needed for server verification and the user is required to type what they chose as the token when they specify the webhook URL. Also, each notification in Facebook can be specifically configured, so there might be instances where users may want to specify different tokens for different kinds of notifications. I don't think we should take away a choice that seems to be such a recurring part of the setup process on Facebook's end. I would love to hear your further thoughts about this. :)

#### About Beeminder:

Okay, so there are three ways to use the webhook:
1. For notifications via PM, in which case a stream need not be created or specified. Only the user's Zulip email suffices.
2. For notifications to a public stream, in which case the user's Zulip email need not be specified, the target stream is enough.
3. For notifications via PMs and streams, depending on which "reminders" are "private" or not. In this case, both the user's Zulip email and a target stream would be specified in the URL and private reminder notifications would be sent as a PM and public reminder notifications would be sent to the target stream.

For now, I divided the docs into two sections, one for PMs, one for streams. This separation seemed justified and simplified the steps and allowed the language and style to not deviate from the style we are aiming for here. However, I haven't addressed the ability to receive both PMs and stream notifications, and I would appreciate your thoughts on whether we should or not. Off the top of my head, users might want either but not both. Not to mention the third point leaves the possibility that a private reminder (such as about the user's weight goals, which can be a personal topic) may inadvertently get posted to a public stream. What do you think?

#### Separate steps for "create a bot" and "construct URL"

I am not so sure about this one. Both steps seemed intertwined enough in most cases that they should go together. However, in certain cases, constructing a URL can be complicated enough to warrant its own step (which is the case with Beeminder, depending on whether the user wants PM or stream notifications). So, I am a bit unsure as to what's the best choice here in the long run and I would appreciate your thoughts! 

As an aside, separating the two would involve a larger refactor in the macros that probably shouldn't be a part of this PR (for example, splitting up the `create-bot-construct-url.md` type macros).

#### Style guide

Here is [the Paper Doc](https://paper.dropbox.com/doc/Style-guide-Documenting-your-webhook-integration-9v28shu2hWnhSnU0oRhdC?_tk=share_copylink) where I started writing the style guide. It is only a first draft, so probably needs lots of improvements. :) Please feel free to edit as you please!

That's all I have for now! Thanks for your guidance, @rishig and @timabbott! :)

\-Eeshan

rishig commented:
Facebook: Let's move this to czo, but it's worth thinking about how we do server verification and notification configuration for our other integrations, that don't have such a token.

Beeminder: hm, let's also take this to czo then, and likely leave it out of the first 5 we merge. It seems complicated in a couple of different ways (e.g. why does "Click on **Make this goal private**" matter?)

Separate steps: sounds good; causing a big refactor is a great reason to leave it as is :). The rule can be "for straightforward things where a macro applies, use a macro". And for everything else, can be a judgement call based on how complicated the URL construction is, with the default being that it's a separate step. 

Also, if you notice that there is a pattern of constructing URLs that is repeated a couple of times, we can consider making a new macro for it.

Style guide: awesome! Looks great; I'll do a quick editing pass tonight or tomorrow, but the rest of the PR can proceed in parallel.

Maybe to make things simple, could you open a separate PR with a couple of the most straightforward integrations from this PR, so we can merge them and reduce the size of this PR?

Thanks @eeshangarg!

eeshangarg commented:
@rishig: Done! This PR now contains only the commits that need more work, the rest have been moved over to #8501! :) Also, this PR will have merge conflicts when #8501 is merged, since commits in both PRs depend on commits that will be merged as part of #8501. I'll take care of that! :)

Renaming this as a WIP till we decide how to move further! Thanks! :)

eeshangarg commented:
@rishig: I just rebased after #8501 was merged and this PR is now as concise as it can possibly be. Here are few things we need to sort out before I go ahead and migrate all webhook docs:

* Finalise the style guide.
* Facebook: we still need to discuss how to approach this one.
* Beeminder: we still need to discuss how to approach this one.
* Crashlytics: could you please see my commit description and share your thoughts on whether we should actually sign up and link an app to a Crashlytics account to verify the correctness of our instructions?
* CircleCI: you mentioned this one isn't complete. I'm not sure, I think it is. What other steps do you think we need? Once the URL to query is in the CircleCI config file, shouldn't CircleCI take care of the rest?
* Dropbox: Could you please share your thoughts on step 3? Should it be split up? There are 4 sentences but I don't see an eloquent way to join the middle ones with an "and", like you suggested.

Thanks!
\-Eeshan

rishig commented:
Great, thanks! For facebook and beeminder, let's start a discussion on czo (maybe in #integrations)?

Crashlytics: sounds good, agreed it is too much work to verify. I would change "the URL we created above" to "the URL created above"

CircleCI: I think we want complete instructions in our documentation, rather than linking to their docs.
E.g. a problem with just linking to docs: "CircleCI also supports Slack, HipChat, Campfire, Flowdock and IRC notifications; you configure these notifications from your project’s Project Settings > Notifications page." Does this apply to zulip or not?
Their documentation is also written for highly technical people, e.g. "This example will POST a JSON packet to the specified URL".

Dropbox: "Dropbox My apps page" -> "Dropbox apps page". Otherwise looks great!

eeshangarg commented:
@rishig: Updated! A few thoughts:
* CircleCI: So, you can't setup a webhook for Zulip as you would for Slack, HipChat, etc. There is no UI you can go to. Even in **Project Settings** --> **Webhook Notifications**, CircleCI simply links to the same place we linked to.
* Facebook: Turns out `token` can't be blank. So I set it to `unused`.
* Beeminder: Dropped the commit for now. Let's resolve the discussion on czo before we move forward!

Looking forward to your feedback! Thanks! :)

rishig commented:
Thanks @eeshangarg!
* CircleCI: Interesting! In any case, it's fine if they have to manually change the config.yml file, but we should give them the instructions for what to change in our documentation, rather than sending them to the circle ci site (particularly because the circle ci documentation is written for a much more technical audience).

Facebook, Beeminder: sounds good! The beeminder discussion is at https://chat.zulip.org/#narrow/stream/127-integrations/topic/incoming.20to.20PMs.

rishig commented:
Thanks @eeshangarg! I merged Dropbox, and Crashlytics after a minor change. 
I also pushed updates to CircleCI and Facebook to https://github.com/rishig/zulip/tree/review-8433

* CircleCI: I think putting in the URL is a bit confusing, so I want to replace it with something like <URL created above>. That causes the formatting bug below; any ideas as to what is going on, or for a workaround?
![image](https://user-images.githubusercontent.com/890911/37057938-d716d80a-213d-11e8-863e-0e7dc19133c9.png)

* Facebook: ongoing discussion in chat.

eeshangarg commented:
@rishig: Updated! Not the best at CSS, so there is probably a better fix for the code-block problem than I what I came up with! Thanks! :)

rishig commented:
Tim, could you take a look at the CSS commit when you get a chance?
The facebook one is still blocked on us understanding what it does, but the CircleCI one can be merged.

eeshangarg commented:
@timabbott, @rishig: Just updated this! Could you please take a look? Thanks! :)

eeshangarg commented:
FYI I don't think this PR has anything to do with the build failure

rishig commented:
Tim, passing to you for reviewing/merging this.

timabbott commented:
I'm going to get Brock's feedback on the CSS change.

timabbott commented:
Sounds good; merged, thanks @eeshangarg!
